### PR TITLE
Masked input

### LINKED ISSUE #4422: Vault UI and a way to obfuscate the secrets
Hi everybody,
We have just started working with Vault, with the File backend, but i've ran into a situation in which i need to hide/obfuscate secrets when i enter in a key. Like another button just to show me the password.
Is there a way to do this with Vault UI? 

Best regards,
Matei

**Feature Request:**
--> Obfuscation

Vault v0.10.0
* Operating System/Architecture: Linux

### PR DESCRIPTION
**Pull request by madalynrose**
Fixes #4422 
screen shot 2018-06-13 at 4 03 32 pm
screen shot 2018-06-13 at 4 03 45 pm
screen shot 2018-06-13 at 4 03 57 pm
screen shot 2018-06-13 at 4 04 21 pm

### CODE REVIEW

File: ui/app/components/masked-input.js [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+export default Ember.Component.extend({
+	value: null,
+	didInsertElement(){
+		this._super(...arguments);
+		autosize(this.element.querySelector('textarea'));
+	},
+	didRender(){
+		this._super(...arguments);
+		autosize.update(this.element.querySelector('textarea'));
+	},
```
Comment (alisdair) [inline ui/app/components/masked-input.js]: The `didRender` hook is also fired on initial render. Maybe you want `didUpdate` here? [Component lifecycle hooks docs here](https://guides.emberjs.com/release/components/the-component-lifecycle/)

File: ui/app/components/masked-input.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		}
+	}),
+	isMasked: true,
+	isFocused: false,
+	displayOnly: false,
+	onKeyDown(){},
+	onChange(){},
+	updateValue(e){
+		this.set('value', e.target.value);
+		this.get('onChange')();
```
Comment (alisdair) [inline ui/app/components/masked-input.js]: This could be written as `this.onChange()`, which might be a little easier to read.

File: ui/app/templates/components/masked-input.hbs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	{{#if displayOnly}}
+		<pre class="masked-value display-only">{{displayValue}}</pre>
+	{{else}}
+  <textarea
+    class="input masked-value"
+    rows=1
+    wrap="off"
+    placeholder="value"
+    onfocus={{action "setFocus" true}}
+    onblur={{action "setFocus" false}}
```
Comment (alisdair) [inline ui/app/templates/components/masked-input.hbs]: I think these could be `{{action (mut setFocus) true}}` and `…false}}`, removing the need for the action.

File: ui/app/templates/components/masked-input.hbs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    onkeydown={{action onKeyDown}}
+    onchange={{action updateValue}}
```
Comment (alisdair) [inline ui/app/templates/components/masked-input.hbs]: I think the `action` helpers here are unneeded, as these attributes are already bound closure actions.

File: ui/tests/pages/components/masked-input.js [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import { attribute, clickable, fillable, focusable, blurrable } from 'ember-cli-page-object';
+
+export default {
+  wrapperClass: attribute('class','[data-test-masked-input]'),
+  enterText: fillable('[data-test-textarea]'),
+  toggleMasked: clickable('[data-test-button]'),
+  focus: focusable('[data-test-textarea]'),
+  blur: blurrable('[data-test-textarea]')
```
Comment (alisdair) [inline ui/tests/pages/components/masked-input.js]: I did not know that `focusable` and `blurrable` were ecpo functions! Super useful.

File: ui/tests/acceptance/console-test.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    });
+  });
+  andThen(() => {
+    enginesPage.console.consoleInput('refresh');
+    enginesPage.console.enter();
+  });
+  andThen(() => {
+    assert.equal(enginesPage.rows().count, numEngines + 3, 'new engines were added to the page');
+  });
+});
```
Comment (alisdair) [inline ui/tests/acceptance/console-test.js]: This test seems unrelated to the rest of the changes. Am I missing something?

Comment (meirish) [reply ui/tests/acceptance/console-test.js]: 🤔 something odd going on - this is already in master https://github.com/hashicorp/vault/commits/master/ui/tests/acceptance/console-test.js

Maybe just needs a rebase? Or was rebased oddly? Uncertain at this moment.

**Comment by joshuaogle** [comment]
Can we make the toggle icon flush with the left column header? In secrets it's just the column header that is flush, but in other places the non-secret text might look odd with it indented

image

File: ui/app/styles/components/masked-input.scss [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+.masked-input .input:focus + .masked-input-toggle {
+	background: rgba($white, 0.95);
+}
+
+.masked-input.masked .masked-value,
+.masked-input.masked .masked-input-toggle {
+	color: $grey-light;
+}
```
Comment (joshuaogle) [inline ui/app/styles/components/masked-input.scss]: Some strange indenting going on in this file

File: ui/app/styles/components/masked-input.scss [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+	line-height: 1.5;
+	font-size: 1rem;
+}
+
+.masked-input-toggle {
+	background: transparent;
+	position: absolute;
+	height: auto;
+	top: $size-6/4;
+	bottom: $size-6/4;
```
Comment (joshuaogle) [inline ui/app/styles/components/masked-input.scss]: Looking forward to the new spacing variables :\

File: ui/app/styles/components/masked-input.scss (hunk 7 above)
Comment (alisdair) [reply ui/app/styles/components/masked-input.scss]: Hard tabs! 🙀

File: ui/app/styles/components/masked-input.scss (hunk 8 above)
Comment (meirish) [reply ui/app/styles/components/masked-input.scss]: Yeah 😬

File: ui/app/styles/components/masked-input.scss (hunk 7 above)
Comment (madalynrose) [reply ui/app/styles/components/masked-input.scss]: Default editor settings! Sorryyyyyy fix'd

**Comment by KalpeshChaniyara** [comment]
Hi @madalynrose ,
We are planning to use Hashicorp Vault along with its ui console, 
I see you have fixed the issue of secret values being displayed in plain text.
I am using binaries of tag v0.10.3
Could you help me with configuration to enabled masked input?
